aten/src/THC/THCCachingAllocator.cpp

```diff
@@ -587,3 +587,50 @@ THC_API uint64_t THCCachingAllocator_maxMemoryCached(int device) {
   assertValidDevice(device);
   return caching_allocator.get_stats_for_device(device).max_amount_cached;
 }
+
+//
+// In CUDA IPC, sender sends a tensor to receiver, THCCaching_CUDAIpcDevptr
+// is called by receiver process to get access to the memory where the tensor
+// was built on sender process.
+//
+// CUDA IPC only allows sharing a big memory block associated with a IpcMemHandle,
+// and it can be opened only **once** per context per process. There can be
+// multiple types of storage in the same IPC mem block, so we must cache the
+// device ptr to construct typed storage as it comes.
+//
+// ipcMemHandle_to_devptr maps ipcMemHandle to a device pointer in the process
+// that can be used to access the memory block in the sender process.
+// It only saves a weak_ptr of the device pointer in the map, the shared_ptr
+// will be used to reconstruct all storages in this memory region. And it will deleted in
+// cudaIpcCloseMemHandle when ref_count is 0.
+//
+namespace {
+  std::mutex IpcMutex;
+  std::unordered_map<std::string, std::weak_ptr<void>> ipcMemHandle_to_devptr;
+}
+
+THC_API std::shared_ptr<void> THCCaching_CUDAIpcDevptr(std::string handle) {
+  std::lock_guard<std::mutex> guard(IpcMutex);
+
+  // If this ipcMemHandle hasn't been opened, or already expired, open it to
+  // enable IPC access to that mem block.
+  if (ipcMemHandle_to_devptr.find(handle) == ipcMemHandle_to_devptr.end()
+      || ipcMemHandle_to_devptr[handle].expired()) {
+    void *devPtr = nullptr;
+    cudaIpcMemHandle_t ipc_handle = *(cudaIpcMemHandle_t*)handle.c_str();
+    THCudaCheck(cudaIpcOpenMemHandle(&devPtr, ipc_handle, cudaIpcMemLazyEnablePeerAccess));
+    // devPtr has to be deleted in same device when created.
+    int curr_device;
+    THCudaCheck(cudaGetDevice(&curr_device));
+    auto sp = std::shared_ptr<void>(
+        devPtr,
+        [curr_device](void *ptr) {
+          THCudaCheck(cudaSetDevice(curr_device));
+          THCudaCheck(cudaIpcCloseMemHandle(ptr));});
```

Comment:
This deleter doesn't look like it actually it's removing entries from the table. So it seems to me that we are leaking memory, because `ipcMemHandle_to_devptr` will just gradually get larger, without ever being GC'ed.